packages/fhir-router/src/graphql/graphql.ts

```diff
@@ -442,25 +444,76 @@ const DEFAULT_MAX_DEPTH = 12;
  */
 const MaxDepthRule =
   (maxDepth: number = DEFAULT_MAX_DEPTH) =>
-  (context: ValidationContext): ASTVisitor => ({
-    Field(
-      /** The current node being visiting. */
-      node: any,
-      /** The index or key to this node from the parent node or Array. */
-      _key: string | number | undefined,
-      /** The parent immediately above this node, which may be an Array. */
-      _parent: ASTNode | readonly ASTNode[] | undefined,
-      /** The key path to get to this node from the root node. */
-      path: readonly (string | number)[]
-    ): any {
-      const depth = getDepth(path);
-      if (depth > maxDepth) {
-        const fieldName = node.name.value;
-        context.reportError(
-          new GraphQLError(`Field "${fieldName}" exceeds max depth (depth=${depth}, max=${maxDepth})`, {
-            nodes: node,
-          })
-        );
+  (context: ValidationContext): ASTVisitor =>
+    new MaxDepthVisitor(context, maxDepth);
+
+type DepthRecord = { depth: number; node?: FieldNode };
+
+class MaxDepthVisitor {
```

Comment:
Should we add `implements ASTVisitor`?

Reply:
I actually tried that: it doesn't work because of the way the `ASTVisitor` type is defined in `graphql-js`